keel-echo/src/test/kotlin/com/netflix/spinnaker/keel/echo/UnhealthyNotificationListenerTests.kt

```diff
@@ -0,0 +1,112 @@
+import com.netflix.spinnaker.keel.events.NotificationEvent
+import com.netflix.spinnaker.keel.events.ResourceHealthEvent
+import com.netflix.spinnaker.keel.notifications.NotificationScope.RESOURCE
+import com.netflix.spinnaker.keel.notifications.NotificationType.UNHEALTHY_RESOURCE
+import com.netflix.spinnaker.keel.persistence.UnhealthyRepository
+import com.netflix.spinnaker.keel.test.locatableResource
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.core.env.Environment
+import java.time.Duration
+
+class UnhealthyNotificationListenerTests : JUnit5Minutests {
+  class Fixture {
+    val config = UnhealthyNotificationConfig()
+    val unhealthyRepository: UnhealthyRepository = mockk(relaxed = true)
+    val springEnv: Environment = mockk(relaxed = true) {
+      every { getProperty("keel.notifications.unhealthy", Boolean::class.java, true)} returns true
+    }
+    val publisher: ApplicationEventPublisher = mockk(relaxed = true)
+    val subject = UnhealthyNotificationListener(
+      config,
+      unhealthyRepository,
+      publisher,
+      springEnv,
+      "https://wwwblah.com"
+    )
+
+    val r = locatableResource()
+    val env = com.netflix.spinnaker.keel.api.Environment(
+      name = "test",
+      resources = setOf(r),
+      notifications = setOf(
+        NotificationConfig(type = NotificationType.slack, address = "#ohmy", frequency = NotificationFrequency.quiet),
+        NotificationConfig(type = NotificationType.email, address = "oh@my.com", frequency = NotificationFrequency.quiet)
+      ),
+      constraints = setOf()
+    )
+  }
+
+  fun tests() = rootContext<Fixture> {
+    fixture {
+      Fixture()
+    }
+
+    context("no existing unhealthy record") {
+      before {
+        every {
+          unhealthyRepository.markUnhealthy(r.id)
+        } returns Duration.ofMinutes(0)
+      }
+
+      test("marks unhealthy") {
+        subject.onResourceHealthEvent(ResourceHealthEvent(resource = r, healthy = true))
+        verify(exactly = 1) { unhealthyRepository.markHealthy(r.id) }
+        verify(exactly = 1) { publisher.publishEvent(ClearNotificationEvent(RESOURCE, r.id, UNHEALTHY_RESOURCE)) }
+      }
+
+      test("marks unhealthy") {
+        subject.onResourceHealthEvent(ResourceHealthEvent(resource = r, healthy = false))
+        verify(exactly = 1) { unhealthyRepository.markUnhealthy(r.id) }
+        verify(exactly = 0) { publisher.publishEvent(any()) }
+      }
+    }
+
+    context("existing unhealthy record") {
+      before {
+        subject.onResourceHealthEvent(ResourceHealthEvent(resource = r, healthy = false))
+      }
+
+      context("it's been 4 minutes (less than the min time)") {
+        before {
+          every {
+            unhealthyRepository.markUnhealthy(r.id)
```

Comment:
I was expecting `durationUnhealthy()` to be called here. This is a little confusing, as I pointed out in the interface, above.